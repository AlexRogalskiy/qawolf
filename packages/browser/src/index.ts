export { Page as PuppeteerPage } from "puppeteer";
export * from "./browser";
export * from "./find";
export * from "./retry";
export * from "./strokes";
