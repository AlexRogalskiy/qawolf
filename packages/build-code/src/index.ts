export * from "./buildInitialCode";
export { stepToSelector } from "./buildSelector";
export * from "./buildVirtualCode";
export * from "./CodeReconciler";
