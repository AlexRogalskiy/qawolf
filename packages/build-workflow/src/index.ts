// export * from "./buildClickSteps";
// export * from "./buildScrollSteps";
// export * from "./buildSelectSteps";
// export * from "./buildSteps";
// export * from "./buildTypeSteps";
// export * from "./buildWorkflow";
