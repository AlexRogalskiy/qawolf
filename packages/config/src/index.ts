import dotenv from "dotenv";

const dotEnvPath = process.env.DOTENV_PATH;
dotenv.config(dotEnvPath ? { path: dotEnvPath } : {});

const testUrl = process.env.TEST_URL || "http://localhost:5000/";

const parseBool = (value: string | undefined) => {
  return !!value && value !== "false";
};

const parseNumber = (value: string | undefined) => {
  if (typeof value === "undefined") return value;

  return parseFloat(value);
};

export const CONFIG = {
  chromeExecutablePath: process.env.CHROME_EXECUTABLE_PATH,
  dataAttribute: process.env.DATA_ATTRIBUTE || null,
  headless: parseBool(process.env.HEADLESS),
  keepBrowserOpen: parseBool(process.env.KEEP_BROWSER_OPEN),
  logPath: process.env.LOG_PATH,
  saveBrowserWsPath: process.env.SAVE_BROWSER_WS_PATH,
  screenshotPath: process.env.SCREENSHOT_PATH,
  sleepAfterEach: parseNumber(process.env.SLEEP_AFTER_EACH),
  testUrl,
  useLocalModule: parseBool(process.env.USE_LOCAL_MODULE)
};
