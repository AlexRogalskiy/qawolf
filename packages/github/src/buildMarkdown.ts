import {
  AggregatedResult,
  AssertionResult,
  TestResult
} from "@jest/test-result";

export const buildMarkdown = (results: AggregatedResult): string => {
  const markdown: string[] = [];

  if (results.numFailedTestSuites) {
    markdown.push(`<h2>${results.numFailedTests} Failed</h2>`);
  }

  results.testResults.forEach(testResult => {
    if (testResult.numFailingTests) {
      markdown.push(buildWorkflowMarkdown(testResult));
    }
  });

  if (results.numPassedTestSuites) {
    markdown.push(`<h2>${results.numPassedTestSuites} Passed</h2>`);
  }

  results.testResults.forEach(testResult => {
    if (!testResult.numFailingTests) {
      markdown.push(buildWorkflowMarkdown(testResult));
    }
  });

  return markdown.join("");
};

const buildStepsMarkdown = (assertionResults: AssertionResult[]): string => {
  const markdown: string[] = ["<br /><b>Steps</b><br /><ol>"];

  assertionResults.forEach(result => {
    const emoji = result.status === "passed" ? ":white_check_mark:" : ":x:";
    const stepName =
      result.status === "pending" ? `<s>${result.title}</s>` : result.title;

    markdown.push(`<li>${emoji} ${stepName}</li>`);
  });

  markdown.push("</ol>");

  return markdown.join("");
};

const buildWorkflowMarkdown = (testResult: TestResult): string => {
  const emoji = testResult.numFailingTests ? ":x:" : ":white_check_mark:";

  const testName = getTestNameFromFile(testResult.testFilePath);
  const steps = buildStepsMarkdown(testResult.testResults);

  return `<details><summary>${emoji} ${testName}</summary>${steps}</details>`;
};

const getTestNameFromFile = (testFilePath: string): string => {
  const splitFilePath = testFilePath.split("/");
  const fileName = splitFilePath[splitFilePath.length - 1];
  const fileNameWithoutExtension = fileName.split(".test")[0];

  return fileNameWithoutExtension;
};
