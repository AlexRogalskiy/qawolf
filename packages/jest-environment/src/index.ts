import { RunnerEnvironment } from "./RunnerEnvironment";
import "./types";

module.exports = RunnerEnvironment;
