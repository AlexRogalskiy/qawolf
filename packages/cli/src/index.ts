#!/usr/bin/env node

import { buildTest } from "@qawolf/build-test";
// import { buildWorkflow } from "@qawolf/build-workflow";
import { Browser } from "@qawolf/browser";
import { CONFIG } from "@qawolf/config";
import { logger } from "@qawolf/logger";
import { Runner } from "@qawolf/runner";
import { Workflow } from "@qawolf/types";
import program from "commander";
import { outputFile, outputJson, readJson } from "fs-extra";
import { snakeCase } from "lodash";
import { resolve } from "path";
import { runTest } from "./runTest";
import { parseUrl, getUrlRoot } from "./utils";

// program
//   .command("build <eventsPath> <name>")
//   .description("build a test from events")
//   .action(async (eventsPath, name) => {
//     const sourcePath = resolve(eventsPath);
//     logger.verbose(`read events from ${sourcePath}`);
//     const events = await readJson(sourcePath);

//     const destPath = `${process.cwd()}/.qawolf`;
//     const formattedName = snakeCase(name);
//     const destWorkflowPath = `${destPath}/workflows/${formattedName}.json`;
//     const destTestPath = `${destPath}/tests/${formattedName}.test.js`;

//     logger.verbose(`build workflow -> ${destTestPath}`);
//     const workflow = buildWorkflow(events, formattedName);
//     await outputJson(destWorkflowPath, workflow, { spaces: " " });

//     logger.verbose(`build test -> ${destTestPath}`);
//     const test = buildTest(workflow);
//     await outputFile(destTestPath, test, "utf8");

//     process.exit(0);
//   });

let recordCommand = program
  .command("record <url> [name]")
  .description("record a workflow and create a test");

if (CONFIG.development) {
  recordCommand = recordCommand.option(
    "-e, --events",
    "record events (for debugging)"
  );
}

recordCommand.action(async (urlArgument, optionalName, cmd) => {
  const url = parseUrl(urlArgument);
  const name = optionalName || getUrlRoot(url);
  logger.verbose(`record url "${url.href}"`);

  const browser = await Browser.create({ record: true, url: url.href });
  await browser.waitForClose();

  const destPath = `${process.cwd()}/.qawolf`;
  const formattedName = snakeCase(name);

  if (cmd.events) {
    const destEventsPath = `${destPath}/events/${formattedName}.json`;
    logger.verbose(`create events "${name}" -> ${destEventsPath}`);
    await outputJson(destEventsPath, browser.events, { spaces: " " });
  } else {
    logger.verbose(`create test "${name}"`);
  }

  process.exit(0);
});

program
  .command("run [name]")
  .description("run a workflow")
  .action(async name => {
    const workflowPath = `${process.cwd()}/.qawolf/workflows/${snakeCase(
      name
    )}.json`;
    const workflow = (await readJson(workflowPath)) as Workflow;
    const runner = await Runner.create(workflow);

    await runner.run();
    await runner.close();

    process.exit(0);
  });

program
  .command("test [name]")
  .description("run a test")
  .action(async name => {
    const results = await runTest(name ? snakeCase(name) : null);
    const success = results.numFailedTestSuites < 1;

    process.exit(success ? 0 : 1);
  });

program.allowUnknownOption(false);

program.parse(process.argv);

if (!process.argv.slice(2).length) {
  console.log("\n");
  program.outputHelp();
}
