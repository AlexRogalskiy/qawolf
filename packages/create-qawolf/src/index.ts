#!/usr/bin/env node
import { install as installCi } from 'playwright-ci';
import {
  logError,
  logInstallDependencies,
  logUseTypeScript,
  promptRootDir,
} from './cli';
import { detectTypeScript, writeConfig } from './config';
import {
  addDevDependencies,
  readPackageJson,
  installDependencies,
} from './packageJson';

(async (): Promise<void> => {
  try {
    const useYarn = process.argv[process.argv.length - 1] === '--yarn';

    // create a new line for yarn create
    console.log();

    // run this first to ensure package.json
    await readPackageJson();

    const rootDir = await promptRootDir();

    const useTypeScript = await detectTypeScript();
    logUseTypeScript(useTypeScript);

    await installCi(true);

    await writeConfig({ rootDir, useTypeScript });

    const packages = await addDevDependencies(useTypeScript);

    logInstallDependencies(packages, useYarn);

    installDependencies(useYarn);
  } catch (error) {
    logError(error);
    process.exit(1);
  }
})();
