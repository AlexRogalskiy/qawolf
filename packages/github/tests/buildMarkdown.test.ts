import { buildMarkdown } from "../src/buildMarkdown";
import {
  jestResults,
  jestResultsSingleFailed,
  jestResultsSinglePassed
} from "../fixtures/jestResults";

describe("buildMarkdown", () => {
  it("returns correct markdown for test suite with single failed test", async () => {
    const markdown = buildMarkdown(jestResultsSingleFailed);

    expect(markdown).toMatchSnapshot();
  });

  it("returns correct markdown for test suite with single passed test", async () => {
    const markdown = buildMarkdown(jestResultsSinglePassed);

    expect(markdown).toMatchSnapshot();
  });

  it("returns correct markdown for test suite with multiple tests", async () => {
    const markdown = buildMarkdown(jestResults);

    expect(markdown).toMatchSnapshot();
  });
});
