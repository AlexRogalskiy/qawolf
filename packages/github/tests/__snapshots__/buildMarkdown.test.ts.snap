// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`buildMarkdown returns correct markdown for test suite with multiple tests 1`] = `"<h2>1 Failed</h2><details><summary>:x: failure</summary><br /><b>Steps</b><br /><ol><li>:white_check_mark: step 1</li><li>:white_check_mark: step 2</li><li>:x: step 3</li><li>:x: <s>step 4</s></li></ol></details><h2>1 Passed</h2><details><summary>:white_check_mark: success</summary><br /><b>Steps</b><br /><ol><li>:white_check_mark: step 1</li><li>:white_check_mark: step 2</li><li>:white_check_mark: step 3</li></ol></details>"`;

exports[`buildMarkdown returns correct markdown for test suite with single failed test 1`] = `"<h2>1 Failed</h2><details><summary>:x: failure</summary><br /><b>Steps</b><br /><ol><li>:white_check_mark: step 1</li><li>:white_check_mark: step 2</li><li>:x: step 3</li><li>:x: <s>step 4</s></li></ol></details>"`;

exports[`buildMarkdown returns correct markdown for test suite with single passed test 1`] = `"<h2>1 Passed</h2><details><summary>:white_check_mark: success</summary><br /><b>Steps</b><br /><ol><li>:white_check_mark: step 1</li><li>:white_check_mark: step 2</li><li>:white_check_mark: step 3</li></ol></details>"`;
